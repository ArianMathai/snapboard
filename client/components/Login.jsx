import React, { useMemo, useState} from "react";
import {Link, useNavigate} from "react-router-dom";

function Login() {

    const [username, setUsername] = useState("");
    const [password, setPassword] = useState("");
    const [errorMessage,setErrorMessage] = useState();

    const navigate = useNavigate()

    const isValidCredentials = useMemo(() => {
        return username.length === 0 && password.length === 0;
    }, [ username, password])


    async function handleLogin(e) {
        e.preventDefault();

        const res = await fetch("/api/authentication/login", {
            method: "post",
            body: JSON.stringify({ username, password }),
            headers: {
                "content-type": "application/json",
            },
        });
        if (res.ok) {
            setUsername("");
            setPassword("");
            navigate("/snapboard");
        }

        const data = await res.json();

        setErrorMessage(data.message)
        setUsername("");
        setPassword("");
    }


    return (
        <div>
            <form>
                <label>
                    Username:
                    <div>
                    <input onChange={(e) => setUsername(e.target.value)} type="text" name="username" />
                    </div>
                </label>
                <label>
                    Password:
                    <div>
                    <input onChange={(e) => setPassword(e.target.value)} type="password" name="password" />
                    </div>
                </label>
                <div>
                    <button type="button" onClick={(e) => (handleLogin(e))} disabled={isValidCredentials}>
                        Sign in
                    </button>
                    <Link to="/signup">
                        <button>Sign up</button>
                    </Link>
                    {errorMessage?<div>{errorMessage}</div>:null}
                </div>
            </form>
        </div>
    );
}

export default Login;
