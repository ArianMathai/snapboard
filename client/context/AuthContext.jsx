import React, {createContext, useContext, useState} from 'react';

const AuthContext = createContext(undefined, undefined);

export function useAuth(){
    return useContext(AuthContext);
}

export const AuthProvider = ({ children }) => {
    const [auth, setAuth] = useState(false);

    return (
        <AuthContext.Provider value={{ auth, setAuth }}>
            {children}
        </AuthContext.Provider>
    );
};