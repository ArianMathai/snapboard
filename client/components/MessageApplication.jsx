import {Link} from "react-router-dom";
import MessageApplicationRoutes from "./MessageApplicationRoutes";
import {useAuth} from "../context/AuthContext";

function MessageApplication (){

    const {isAuthorized, setIsAuthorized} = useAuth();

    const handleLogout = () => {
        document.cookie = "authorization=; expires=Thu, 01 Jan 1970 00:00:00 UTC; path=/;";
        setIsAuthorized(false);
    };


    return (
    <>
        <header>
            <Link to="/" style={{ textDecoration: 'none', color: 'black' }}>
                <h2>SnapBoard üëª</h2>
            </Link>
        </header>
        <nav>
            <Link to={"/"}>Home</Link>
            <Link to={"/snapboard"}>SnapBoard</Link>
            <div style={{flex: 1}}></div>
            {!isAuthorized?<Link to={"/login"}>Login</Link>
                : <Link to={"/login"} onClick={handleLogout}>Logout</Link>}
        </nav>
        <main>
            <MessageApplicationRoutes/>
        </main>
        <footer>By Arian‚ò†Ô∏è and Simen‚ò†Ô∏è</footer>
    </>
    );
}
export default MessageApplication;