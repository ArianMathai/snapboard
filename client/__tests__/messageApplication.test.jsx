import MessageApplication from "../components/MessageApplication";
import renderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";
import {AuthProvider, useAuth} from "../context/AuthContext";
import Login from "../components/Login";

describe("Message Application", () => {
    it("renders correctly", () => {

        const component = renderer.create(
            <AuthProvider>
            <MemoryRouter>
                <MessageApplication />
            </MemoryRouter>
            </AuthProvider>
        );
        // Use this line to create or update a snapshot
        expect(component).toMatchSnapshot();
    });
});

//Should we have all our snapshots in the same test file?
describe ("Sign in page renders correctly", () => {
    it("renders correctly", () =>{

        const component = renderer.create(
            <AuthProvider>
                <MemoryRouter>
                    <Login></Login>
                </MemoryRouter>
            </AuthProvider>
        )
        expect(component).toMatchSnapshot();
    })
} )



