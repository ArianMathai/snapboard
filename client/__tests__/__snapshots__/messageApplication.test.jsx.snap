// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message Application renders correctly 1`] = `
[
  <header>
    <a
      href="/"
      onClick={[Function]}
      style={
        {
          "color": "black",
          "textDecoration": "none",
        }
      }
    >
      <h2>
        SnapBoard üëª
      </h2>
    </a>
  </header>,
  <nav>
    <a
      href="/"
      onClick={[Function]}
    >
      Home
    </a>
    <a
      href="/snapboard"
      onClick={[Function]}
    >
      SnapBoard
    </a>
    <div
      style={
        {
          "flex": 1,
        }
      }
    />
    <a
      href="/login"
      onClick={[Function]}
    >
      Login
    </a>
  </nav>,
  <main>
    <div
      className="home-container"
    >
      <p
        className="homeText"
      >
        Velkommen til det digitale anarkiet, et fristed for dine tanker og meninger hvor ingen ansiktstrekk avsl√∏res. Dette er din virtuelle maskerade, en helligdom for fri tanke. Her kan du kaste din stemme til vinden og la dine ord danse anonymt blant de digitale skyggene. Respekt er n√∏kkelen, selv i anonymitetens skjul, for vi tror p√• den usynlige sosiale kontrakten som binder oss alle sammen. Bli en del av dette sanatoriumet for det usagte, men husk, selv i anonymitetens frihet, har ordene dine makt. Tr√• varsomt i denne digitale villmarken.Velkommen til ville vesten.
      </p>
    </div>
  </main>,
  <footer>
    By Arian‚ò†Ô∏è and Simen‚ò†Ô∏è
  </footer>,
]
`;

exports[`Sign in page renders correctly renders correctly 1`] = `
<div>
  <form>
    <label>
      Username:
      <div>
        <input
          name="username"
          onChange={[Function]}
          type="text"
        />
      </div>
    </label>
    <label>
      Password:
      <div>
        <input
          name="password"
          onChange={[Function]}
          type="password"
        />
      </div>
    </label>
    <div>
      <button
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Sign in
      </button>
      <a
        href="/signup"
        onClick={[Function]}
      >
        <button>
          Sign up
        </button>
      </a>
    </div>
  </form>
</div>
`;
